File: php/src/Metrics/HdrEncoder.php
<?php

declare(strict_types=1);

namespace RespBench\Metrics;

use RuntimeException;

/**
 * Encoder for the HdrHistogram V2 compressed binary format, byte-compatible with
 * Java's Histogram.encodeIntoCompressedByteBuffer() (and the Ruby engine's encoder).
 *
 * All integers are big-endian (network byte order).
 *
 * Compressed wrapper:
 *   int32  compressed_encoding_cookie   (0x1c849314)
 *   int32  compressed_payload_length
 *   byte[] zlib-deflated V2 payload
 *
 * V2 payload:
 *   int32  v2_encoding_cookie           (0x1c849313)
 *   int32  payload_length
 *   int32  normalizing_index_offset     (0)
 *   int32  number_of_significant_digits
 *   int64  lowest_trackable_value
 *   int64  highest_trackable_value
 *   int64  integer_to_double_conversion_ratio  (IEEE754 double bits; 1.0)
 *   byte[] counts (ZigZag LEB128, negative = zero-run)
 */
final class HdrEncoder
{
    private const V2_ENCODING_COOKIE_BASE = 0x1c849303;
    private const V2_COMPRESSED_ENCODING_COOKIE_BASE = 0x1c849304;
    private const WORD_SIZE_FLAG = 0x10;

    private const V2_ENCODING_COOKIE = self::V2_ENCODING_COOKIE_BASE | self::WORD_SIZE_FLAG; // 0x1c849313
    private const V2_COMPRESSED_ENCODING_COOKIE = self::V2_COMPRESSED_ENCODING_COOKIE_BASE | self::WORD_SIZE_FLAG; // 0x1c849314

    /**
     * Encode a histogram into the V2 compressed binary format, base64-encoded.
     */
    public static function encodeCompressedBase64(HdrHistogram $histogram): string
    {
        return base64_encode(self::encodeCompressed($histogram));
    }

    public static function encodeCompressed(HdrHistogram $histogram): string
    {
        $v2Payload = self::encodeV2($histogram);
        $compressed = gzcompress($v2Payload, 6);
        if ($compressed === false) {
            throw new RuntimeException('Failed to zlib-compress HDR payload');
        }

        // int32 cookie + int32 compressed length + compressed bytes
        return pack('N2', self::V2_COMPRESSED_ENCODING_COOKIE, strlen($compressed)) . $compressed;
    }

    private static function encodeV2(HdrHistogram $histogram): string
    {
        $normalizingOffset = 0;
        $sigFigs = $histogram->significantFigures;
        $lowest = $histogram->lowestTrackableValue;
        $highest = $histogram->highestTrackableValue;
        $conversionRatioBits = self::doubleToLongBits(1.0);

        $countsBytes = self::encodeCounts($histogram);

        // payload_len = everything after the payload_len field itself:
        // normalizing(4) + sigfigs(4) + lowest(8) + highest(8) + ratio(8) + counts
        $payloadLen = 4 + 4 + 8 + 8 + 8 + strlen($countsBytes);

        $header = pack('N4', self::V2_ENCODING_COOKIE, $payloadLen, $normalizingOffset, $sigFigs);
        $int64Fields = self::packInt64BE($lowest)
            . self::packInt64BE($highest)
            . self::packInt64BE($conversionRatioBits);

        return $header . $int64Fields . $countsBytes;
    }

    private static function encodeCounts(HdrHistogram $histogram): string
    {
        $relevantLength = $histogram->relevantLength();
        $result = '';
        $index = 0;

        while ($index < $relevantLength) {
            $count = $histogram->rawCountAt($index);
            if ($count === 0) {
                $zeros = 1;
                while (($index + $zeros) < $relevantLength && $histogram->rawCountAt($index + $zeros) === 0) {
                    $zeros++;
                }
                $result .= self::encodeZigZag(-$zeros);
                $index += $zeros;
            } else {
                $result .= self::encodeZigZag($count);
                $index++;
            }
        }

        return $result;
    }

    /**
     * ZigZag + LEB128 encode a signed 64-bit integer.
     */
    private static function encodeZigZag(int $value): string
    {
        // ZigZag: (value << 1) ^ (value >> 63), using arithmetic shift semantics.
        $zz = ($value << 1) ^ ($value >> 63);

        $result = '';
        // Process as unsigned 64-bit via logical right shift.
        while (true) {
            if (($zz & ~0x7F) === 0) {
                $result .= chr($zz & 0x7F);
                break;
            }
            $result .= chr(($zz & 0x7F) | 0x80);
            $zz = self::logicalShiftRight($zz, 7);
        }

        return $result;
    }

    /**
     * Logical (unsigned) right shift for 64-bit ints in PHP.
     */
    private static function logicalShiftRight(int $value, int $bits): int
    {
        if ($bits === 0) {
            return $value;
        }

        // Shift then clear the top `bits` sign-extended bits.
        return ($value >> $bits) & (PHP_INT_MAX >> ($bits - 1));
    }

    private static function doubleToLongBits(float $value): int
    {
        // Pack as big-endian double, unpack as signed big-endian int64.
        $packed = pack('E', $value); // 'E' = big-endian double
        /** @var array{1:int} $unpacked */
        $unpacked = unpack('J', $packed); // 'J' = big-endian unsigned int64

        return $unpacked[1];
    }

    private static function packInt64BE(int $value): string
    {
        // 'J' handles the full 64-bit range; PHP ints are 64-bit signed.
        return pack('J', $value);
    }
}

File: go/internal/metrics/hdrencoder.go
package metrics

import (
	"bytes"
	"compress/zlib"
	"encoding/base64"
	"encoding/binary"
	"math"
)

// HdrHistogram V2 encoding constants (Java-compatible).
const (
	v2EncodingCookieBase           = 0x1c849303
	v2CompressedEncodingCookieBase = 0x1c849304
	wordSizeFlag                   = 0x10

	v2EncodingCookie           = v2EncodingCookieBase | wordSizeFlag           // 0x1c849313
	v2CompressedEncodingCookie = v2CompressedEncodingCookieBase | wordSizeFlag // 0x1c849314
)

// EncodeCompressedBase64 encodes a histogram into the V2 compressed binary
// format and base64-encodes it, byte-compatible with Java's
// Histogram.encodeIntoCompressedByteBuffer().
func EncodeCompressedBase64(h *HdrHistogram) (string, error) {
	compressed, err := EncodeCompressed(h)
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(compressed), nil
}

// EncodeCompressed produces the compressed wrapper around the V2 payload.
func EncodeCompressed(h *HdrHistogram) ([]byte, error) {
	payload := encodeV2(h)

	var buf bytes.Buffer
	zw, err := zlib.NewWriterLevel(&buf, zlib.DefaultCompression)
	if err != nil {
		return nil, err
	}
	if _, err := zw.Write(payload); err != nil {
		return nil, err
	}
	if err := zw.Close(); err != nil {
		return nil, err
	}
	compressed := buf.Bytes()

	out := make([]byte, 8+len(compressed))
	binary.BigEndian.PutUint32(out[0:4], uint32(v2CompressedEncodingCookie))
	binary.BigEndian.PutUint32(out[4:8], uint32(len(compressed)))
	copy(out[8:], compressed)
	return out, nil
}

// encodeV2 builds the uncompressed V2 payload.
func encodeV2(h *HdrHistogram) []byte {
	countsBytes := encodeCounts(h)

	// Java's V2 payload_length is the counts-array length ONLY — written as
	// `buffer.position() - payloadStartPosition`, measured after the 40-byte
	// header. It must NOT include the 32 bytes of header fields that follow the
	// length field, or Java's decoder throws "The buffer does not contain the
	// indicated payload amount".
	payloadLen := len(countsBytes)

	buf := make([]byte, 0, 40+len(countsBytes))
	buf = binary.BigEndian.AppendUint32(buf, uint32(v2EncodingCookie))
	buf = binary.BigEndian.AppendUint32(buf, uint32(payloadLen))
	buf = binary.BigEndian.AppendUint32(buf, 0) // normalizing_index_offset
	buf = binary.BigEndian.AppendUint32(buf, uint32(h.SignificantFigures))
	buf = binary.BigEndian.AppendUint64(buf, uint64(h.LowestTrackableValue))
	buf = binary.BigEndian.AppendUint64(buf, uint64(h.HighestTrackableValue))
	buf = binary.BigEndian.AppendUint64(buf, math.Float64bits(1.0)) // conversion ratio
	buf = append(buf, countsBytes...)
	return buf
}

// encodeCounts serializes the counts array using ZigZag LEB128 with zero-run
// compression (negative values encode a run of zeros).
func encodeCounts(h *HdrHistogram) []byte {
	relevant := h.RelevantLength()
	var out []byte
	i := 0
	for i < relevant {
		count := h.RawCountAt(i)
		if count == 0 {
			zeros := 1
			for i+zeros < relevant && h.RawCountAt(i+zeros) == 0 {
				zeros++
			}
			out = appendZigZag(out, int64(-zeros))
			i += zeros
		} else {
			out = appendZigZag(out, count)
			i++
		}
	}
	return out
}

// appendZigZag appends a ZigZag+LEB128-encoded signed 64-bit integer.
func appendZigZag(dst []byte, value int64) []byte {
	// ZigZag transform, then treat as unsigned for LEB128.
	zz := uint64((value << 1) ^ (value >> 63))
	for {
		if zz&^0x7F == 0 {
			return append(dst, byte(zz&0x7F))
		}
		dst = append(dst, byte((zz&0x7F)|0x80))
		zz >>= 7
	}
}
